src/client.rs

```diff
@@ -97,47 +101,94 @@ impl ActivityPubClient {
     }
 
     pub async fn follow_actor(&self, actor_uri: &str) -> Result<()> {
-        let Actor { id, inbox, .. } = self.get_actor(actor_uri).await?;
         let base = &self.base;
-        info!(%id, %inbox, "sending follow request to inbox");
+        let actor: Actor = self.get_actor(actor_uri).await?;
+        let actor_id = actor.id.as_ref().ok_or(Error::StatusAndMessage {
+            status: StatusCode::BAD_REQUEST,
+            message: "actor has no id",
+        })?;
+        let id = actor_id
+            .parse::<http::Uri>()
+            .map_err(|_e| Error::InvalidUri {
+                uri: actor_id.clone(),
+            })?;
+        info!(
+            "sending follow request to inbox: {:?} {:?}",
+            id, actor.inbox
+        );
 
         let message_id = Uuid::new_v4();
-        let message = Activity {
-            context: Context::default(),
-            ty: ActivityType::Follow,
-            to: vec![id.clone()],
-            object: IdOrObject::Id(id),
-            id: format!("https://{base}/activities/{message_id}"),
-            actor: format!("https://{base}/actor)"),
-        };
-
-        self.json_post(inbox, message).await?;
+        let message_id_uri = format!("https://{base}/activities/{message_id}");
+        let actor_uri = format!("https://{base}/actor");
+        let message = ActivityBuilder::new(String::from("Follow"), String::from("Following actor"))
+            .actor(
+                ActorBuilder::new(String::from("Actor")).url(
+                    actor_uri
+                        .parse::<http::Uri>()
+                        .map_err(|_e| Error::InvalidUri { uri: actor_uri })?,
+                ),
+            )
+            .to(vec![actor.id.as_ref().expect("actor has no id").clone()])
```

Comment:
`expect` will panic the server here if the actor has no id. Is that what we want?